text/0025-configuration-policy.md

```diff
@@ -0,0 +1,197 @@
+identical to the same parameters passed into the `Construct` on creation.
+
+All L2s will need to call this hook as part of their object construction. The
+call should be made immediately after the call to `super`, prior to property
+validation or Cfn* resource construction, as the modified properties may alter
+the outcome of these calls.
+
+Having the hook return the (unaltered or modified) properties is necessary
+because many of the FooProps interfaces contain readonly variables
+(e.g., `BucketProps.encryption`).
+
+This approach allows for injecting or enforcing default configuration, even in
+the cases of nested Constructs. For example, when creating a
+DnsValidatedCertificate, the properties of the underlying Lambda Function can
+be altered. However, no special consideration -- for example, an extra
+parameter -- is made to explicitly identify these Constructs as being nested.
+
+# Drawbacks
+
+The largest drawback to this change is the scope of the change -- every L2 will
+need to be extended to call the new hook as part of its initialization. Closely
+related to this, the pattern chosen does not lend itself well to the current
+static enforcement tools available (e.g., `awslint`) to ensure consistent
+behavior across the L2s. A possible mitigation might be to invest in a runtime
+linting capability across the CDK; however, this will expand the scope of the
+implementation considerably.
+
+Given the consistency in the change -- always the same one-liner after
+construction -- it's possible that a simple test can be created to enforce
+that there is one call to the hook per active Construct, simply be examining
+the source. This isn't particularly robust. See Unresolved Questions for more.
+
+# Rationale and Alternatives
+
+Providing this hook is a unique opportunity that provides a way to alter and
+enforce configuration in a way only possible by altering the CDK itself. Other
+similar options for enforcing standard configuration involve factory creation
+methods, sub-classing specific constructs, or hooking into the `onPrepare()`
+(or other existing Construct hooks). Each comes with its own downsides.
+
+Factory methods to create common standard Constructs are one alternate
+solution; this works well enough for top-level Constructs, but falls flat in
+scenarios with "nested" Constructs, or any other scenario where the application
+logic developer isn't directly instantiating the resource. Similar downsides
+exist for end-users subclassing Constructs and initializing them directly.
+
+There are an existing set of hooks for Constructs (e.g., `onPrepare()`) that
+could be used to alter or augment Constructs on creation. However, these hooks
+only receive the initialized Construct without further metadata or properties,
+after baseline Cfn resources have been created and other constructor-time logic
+has been executed. This severely limits the ease and range of configurations
+that can be altered this way.
+
```

Comment:
We should address the use case for "vending" this hook. Example: a team in a company wishes to vend a library of "secure defaults" to other teams. They should be able to vend such a library without needing to vend a base `Stack` class.

One option (which I think is legitimate but might be hard for people to naturally discover) is to simply vend a function:

```ts
import { SecureResources } from '@acme/secure-resources';

class MyClass extends Stack {
  constructor(scope: Construct, id: string) {
    super(scope, id);

    // my resources
  }

  public onResourceCreate(scope: Construct, id: string, props: any) {
    return SecureResources.onResourceCreate(scope, id, props);
  }
}
```

Another alternative is not to define this hook as an override the `Stack` class but use an interface:

```ts
import { SecureResourcesHook } from '@acme/secure-resources'

new Stack(this, 'MyStack', {
  resourceCreationHook: new SecureResourcesHook() // now "creation" makes more sense for some reason
});
```

where `onResourceCreate` is of type:

```ts
interface IResourceCreationHook {
  onCreate(scope: Construct, id: string, props: any): ...
}
```

I would **NOT** be tempted to support multiple hooks since this means that ordering becomes an issue. If people want, they can create a composite hook that invokes other hooks, but then they control the order explicitly.